lib.rs

```diff
@@ -800,8 +810,19 @@ impl<A: Array> SmallVec<A> {
         // from callers like insert()
         let (_, &mut len, cap) = self.triple_mut();
         if cap - len >= additional {
-            return Ok(());
+            Ok(())
+        } else {
+            self.try_reserve_cold(len, additional)
         }
+    }
+
+    #[cold]
+    #[inline(never)]
+    fn try_reserve_cold(
```

Comment:
I'm not really sure about this being tagged as `#[cold]` unconditionally. There's I suspect a somewhat common usage of:

```
let mut vec = SmallVec::new();
vec.reserve(num_elements);

for .... {
  vec.push(..);
}
```

For such cases `#[cold]` is not accurate, and `#[inline(never)]` is probably counter-productive.

I think only the caller knows whether the capacity is going to be likely-already enough. For example `push()` could use the `try_reserve_cold` without much trouble.

Maybe we should expose both, or something...